### PR TITLE
Made optional CGAL dependencies optional.

### PR DESCRIPTION
**Pull request by citibeth**
@adamjstewart 
This PR makes  optional dependencies optional.  It addresses concerns in #2005.  At this point, it has not yet been tested.

### CODE REVIEW

**Comment by adamjstewart** [comment]
As I mentioned in #2005, I would make these variants default to True since the developers listed them as "Essential" dependencies and not "Optional". But I'm not a developer.

**Comment by tgamblin** [comment]
If something requires qt, I think it's reasonable to make that default to `False` mainly due to the expense of building it.  We can reevaluate that when we have binary caching though...

**Comment by tgamblin** [comment]
I think the same holds for `opengl`.  @citibeth: is there any decent reason not to make `core` default to `True`?  I mean, it's called `core`.

**Comment by citibeth** [comment]
> @citibeth <https://github.com/citibeth>: is there any decent reason not
to make core default toTrue? I mean, it's called core.

It's CORE, not a "core" part of CGAL, if that makes any sense:

The directories include/CGAL/CORE and src/CGALCore contain a distribution
> of the Core library[1]
> <http://doc.cgal.org/latest/Manual/installation.html#cite-text-0-0>
> version 1.7 for dealing with algebraic numbers. Core is not part of CGAL
> and has its own license.


http://cs.nyu.edu/exact/core_pages/svn-core.html

The following section from the CGAL manual makes it clear that all these
dependencies are optional.  They are only "essential" if you wish to build
optional parts of CGAL:

[image: Inline image 1]

**Comment by tgamblin** [comment]
Ok.  LGTM then -- I suspect that most people using CGAL in an HPC code are not going to want these three variants on by default.

Still has Travis issues though 😄

File: var/spack/repos/builtin/packages/cgal/package.py [hunk 1]
```diff
@@ -44,13 +44,29 @@ class Cgal(Package):
...
+
+
     # Essential Third Party Libraries
-    depends_on('boost')
+    depends_on('boost+thread+system')
     depends_on('gmp')
     depends_on('mpfr')
+
+    # Required for CGAL_ImageIO
+    depends_on('opengl', when='+imageio')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cgal/package.py]: `opengl` isn't yet in Spack. Although I think `mesa` provides this? That whole mess was very confusing to me.

File: var/spack/repos/builtin/packages/cgal/package.py [hunk 2]
```diff
@@ -44,13 +44,29 @@ class Cgal(Package):
...
     depends_on('zlib')
-    # depends_on('opengl')
-    depends_on('qt@5:')
+
+    # Optional to build CGAL_Qt5 (demos)
+    # depends_on('opengl', when='+demos')   # not yet in Spack
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cgal/package.py]: Is `opengl` required for both demos and imageio?

Comment (citibeth) [inline var/spack/repos/builtin/packages/cgal/package.py]: Yes.

File: var/spack/repos/builtin/packages/cgal/package.py (hunk 1 above)
Comment (citibeth) [inline var/spack/repos/builtin/packages/cgal/package.py]: I decided to not tackle the opengl issue at this time.

Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cgal/package.py]: Can you comment this out for now? Otherwise I imagine `spack install cgal` will crash.

**Comment by citibeth** [comment]
@tgamblin Yes it has flake8 issues.  Flake8 is making me change:
```
        options = std_cmake_args + [
            # CGAL supports only Release and Debug build type. Any
            # other build type will raise an error at configure time
            '-DCMAKE_BUILD_TYPE:STRING=%s' %
                ('Debug' if '+debug' in spec else 'Release'),
            '-DBUILD_SHARED_LIBS:BOOL=%s' %
                ('ON' if '+shared' in spec else 'OFF'),
            '-DWITH_CGAL_Core:BOOL=%s' %
                ('YES' if '+core' in spec else 'NO'),
            '-DWITH_CGAL_ImageIO:BOOL=%s' %
                ('YES' if '+imageio' in spec else 'NO'),
            '-DWITH_CGAL_Qt5:BOOL=%s' %
                ('YES' if '+demos' in spec else 'NO')
        ]
```
into:
```
        options = std_cmake_args + [
            # CGAL supports only Release and Debug build type. Any
            # other build type will raise an error at configure time
            '-DCMAKE_BUILD_TYPE:STRING=%s' %
            ('Debug' if '+debug' in spec else 'Release'),
            '-DBUILD_SHARED_LIBS:BOOL=%s' %
             ('ON' if '+shared' in spec else 'OFF'),
            '-DWITH_CGAL_Core:BOOL=%s' %
            ('YES' if '+core' in spec else 'NO'),
            '-DWITH_CGAL_ImageIO:BOOL=%s' %
            ('YES' if '+imageio' in spec else 'NO'),
            '-DWITH_CGAL_Qt5:BOOL=%s' %
            ('YES' if '+demos' in spec else 'NO')]
```
...which is significantly less readable.

**Comment by tgamblin** [comment]
I don't think either of those is particularly readable.  Sometimes `flake8` inadvertently makes code better 😄.

I'm not breaking `develop` so that this package can have an exception -- I think we've got a reasonable set of rules and they're what most of the Python open source ecosystem uses.  The point isn't to choose the most readable thing -- it's to stop having to argue about style.

**Comment by tgamblin** [comment]
FWIW, I like to do this for stuff like that:

```python
def cm_opt(string, cond, yes='YES', no='NO'):
    return string % (yes if cond in spec else no)

options = std_cmake_args + [
    # CGAL supports only Release and Debug build type. Any
    # other build type will raise an error at configure time
    cm_opt('-DCMAKE_BUILD_TYPE:STRING=%s', '+debug', 'Debug', 'Release'),
    cm_opt('-DCMAKE_BUILD_TYPE:STRING=%s', '+shared', 'ON', 'OFF'),
    cm_opt('-DCMAKE_BUILD_TYPE:STRING=%s', '+shared', 'ON', 'OFF'),
    cm_opt('-DWITH_CGAL_Core:BOOL=%s', '+core'),
    cm_opt('-DWITH_CGAL_ImageIO:BOOL=%s', '+imageio'),
    cm_opt('-DWITH_CGAL_ImageIO:BOOL=%s', '+imageio'),
    cm_opt('-DWITH_CGAL_Qt5:BOOL=%s', '+demos')]
```

**Comment by citibeth** [comment]
I tried that; it didn't save me enough characters to avoid the newline.
Although I suppose the newline would allow for more readability under
flake8 rules.  I didn't think of that at the time.

**Pull request merged**